### PR TITLE
Fix issue 20938 - Cannot create const arrays mixing immutable and mutable structs with indirections

### PR DESCRIPTION
**Pull request by Biotronic**

### CODE REVIEW

**Comment by Geod24** [comment]
Looking at the fix, could you also add a test for this:
```D
class Foo {}
immutable Foo o1;
Object o2;
auto arr = [ o1, o2]; // Common type is `const(Object)`
```

Also please make the commit title match the issue title (add the bug summary, so people using `git log` / `git blame` don't have to look at the diff).

**Comment by Biotronic** [comment]
I'm confused by the errors in circleci - I've not touched any of the things it complains about, and don't really see how I could have caused those errors. On the other hand, other PRs don't seem to have them.

**Comment by UplinkCoder** [comment]
This could use a comment.

**Comment by UplinkCoder** [comment]
@Biotronic The source change alone does not tell me what you did.
I know that you changed the way DMD interprets the semantics of the langauge.
But I don't exactly know in which way.

**Comment by Biotronic** [comment]
@UplinkCoder You're right. Somehow I forgot this thing was a compiler. :p

**Comment by MoonlightSentinel** [comment]
CircleCI does an additional self-compilation with your changes. They apparently cause a previously working overload resolution to fail.

**Comment by UplinkCoder** [comment]
It's a curious thing.
Fixing a legitimate bug can break code.
Since now overloads become ambiguous or `is-expressions` which detected an error before and are part of the meta-programming part of a piece of software now return true and change behavior silently.

**Comment by Geod24** [comment]
> I'm confused by the errors in circleci - I've not touched any of the things it complains about, and don't really see how I could have caused those errors. On the other hand, other PRs don't seem to have them.

Despite what people would like to believe, the CI is very often right.
The failure is:
```
Test fail_compilation/ice13024.d failed.  The logged output:
/home/circleci/dmd/generated/linux/debug/64/dmd -conf= -m64 -Ifail_compilation -verrors=0  -fPIC  -odtest_results/fail_compilation -oftest_results/fail_compilation/ice13024_0.o  -c fail_compilation/ice13024.d 


==============================
Test fail_compilation/ice13024.d failed: expected rc == 1, but exited with rc == 0
```

The test:
https://github.com/dlang/dmd/blob/2b0ce2768d511efe61d3e59cb59d30dea99f79fe/test/fail_compilation/ice13024.d#L1-L16

Now if you look at the error message (L4), you see it expects an error because the user is tryiing to make an array of two different `enum`. And your fix is about allowing array type to be inferred to their common type. Do `enum A` and `enum B` have a type in common ? Yes they do, it's `int`. Hence your fix changes the `auto` deduction from `A[]` to `int[]`. That makes perfect sense to me.

Now is that something we want ? I'd like to hear from other people. On one end, it would be consistent. On the other hand, for `enum` especially, it might be surprising.

In fact, there were two bug reports in the issue: one about not being able to create an array of a common type despite specifying the type, and the other about inference. This is trying to solve the inference problem, which will "hide" the other one.
But the other one is more deep-seated: the D frontend currently has no knowledge of the "expected" type (AFAIK): It just deduce a type and then attempts to convert to the expected one. But that's a deeper, and much more involved issue.

**Comment by Biotronic** [comment]
Yeah, the closest we get is probably polysemous values, and I guess that concept could be expanded to a few other cases, but I don't think this is one of them.

Interestingly, while DMD currently chokes on `[EnumA.Item, EnumB.Item]`, just adding an `int` value in a strategic location makes it compile:

```d
     auto a = [EnumA.Item, 0, EnumB.Item]; // It's a compilation error!
     auto b = [EnumA.Item, EnumB.Item, 0]; // It's an int[]!
```

I would argue `ice13024.d` should be removed and the conversion to `int` be implemented as in this PR. The [documentation](https://dlang.org/spec/expression.html#array_literals) on array literals says "The common type of the all elements is taken to be the type of the array element, and all elements are implicitly converted to that type." I don't think the method for determining the common type is defined in the documentation, but Walter [did originally state](https://forum.dlang.org/post/hdhvgo$u6t$1@digitalmars.com) it should use the logic of the ternary operator to figure it out. No special case is mentioned for enums, and `ice13024.d` isn't really intended to check that `[EnumA.Item, EnumB.Item]` doesn't compile, just that it doesn't [crash the compiler](https://issues.dlang.org/show_bug.cgi?id=13024).

**Comment by Geod24** [comment]
Instead of removing it, just use different base type. E.g. let one as default, and use `string` for another.

**Comment by Biotronic** [comment]
I guess. I tried downloading some old DMD compilers to recreate the issue, with no luck. FWIW, the PR that fixed 13024 [mentions](https://github.com/dlang/dmd/pull/3733#issuecomment-48574896) that the inferred type should be `int[]`.

**Comment by Geod24** [comment]
> I tried downloading some old DMD compilers to recreate the issue, with no luck. 

Did you try 2.066 ?

**Comment by Geod24** [comment]
Side note: Your email address is not tied to your Github profile so the commits are not properly attributed. You might want to fix that.

**Comment by Biotronic** [comment]
Done, thanks. :) And yeah, I did try 2.066 and 2.065 - it compiles fine on both. 13024 is listed as fixed in 2.066's [changelog](https://dlang.org/changelog/2.066.0.html), so I expected the issue to be present in 2.065, but it's not there either. Running it on [run.dlang.org](https://run.dlang.io/is/FHV9vU) also gives no ICE, so I don't know what's happening.

File: test/fail_compilation/ice13024.d [hunk 1]
```diff
@@ -12,5 +12,6 @@ void main()
 {
     T t;
     auto r1 = [cast(int)(t.x), cast(int)(t.y)]; // OK
-    auto r3 = [t.x, t.y]; // crash
+    // Used to be [t.x, t.y]. Behavior changed with fix of issue 20938
```
Comment (Geod24) [inline test/fail_compilation/ice13024.d:15]: What I meant was to change L9 from `enum B { b }` to `enum B : string { b = "42" }`

Comment (Biotronic) [reply test/fail_compilation/ice13024.d:15]: But then L14 wouldn't compile, and much bigger changes would need to be done to the whole test. I'm starting to think `static assert(false, "This test no longer fails to compile after issue 20938 was fixed.");` might be the best solution.

The root problem is the test for 13024 checks more than it should - it should only check if `[t.x, t.y]` crashes the compiler, not that it fails with a specific error message. I don't know how possible that would be, though.

**Comment by UplinkCoder** [comment]
Anything which changes behavior should at least get a revert switch. 
And a dedicated entry in the changelog

File: test/fail_compilation/ice13024.d (hunk 1 above)
Comment (Geod24) [reply test/fail_compilation/ice13024.d:15]: Right, I missed that.
Then I think the best course of action here is to add to `b20938.d` a test with two enums that just checks it compiles, and remove `ice13024.d`.

**Comment by Geod24** [comment]
> Anything which changes behavior should at least get a revert switch.

That's not true. Anything that changes behavior of working code comes with a deprecation. This is fixing an `rejects-valid`, it would be counter-productive to require a revert switch for it.

> And a dedicated entry in the changelog

We provide changelog entries for two cases:
1) User-visible changes without associated bug trackers;
2) Significant changes;

While a nice improvement, we've pulled bigger changes without changelog entry. If @Biotronic is willing to write a couple lines about this it'd be great, but that's not a blocker for the merge.

File: test/fail_compilation/ice13024.d (hunk 1 above)
Comment (Biotronic) [reply test/fail_compilation/ice13024.d]: Done.

**Review by Geod24 — APPROVED** [review]
Awesome work, thanks!

**Comment by UplinkCoder** [comment]
This caused a regression.

https://issues.dlang.org/show_bug.cgi?id=21285

**Pull request merged**